Builders/src/main/scala/Builders.scala

```diff
@@ -0,0 +1,266 @@
+package com.gu.nitf.model.builders
+
+import java.net.URI
+import java.time.{LocalDate, LocalDateTime}
+
+import com.gu.nitf.model._
+import com.gu.nitf.scalaxb._
+import com.gu.scalaxb.{dataRecord => dr}
+import com.gu.scalaxb._
+import scalaxb._
+
+import scala.xml.NodeSeq
+
+
+object `package` {
+  // aliases for builder classes whose names are different from the NITF spec
+  type DocdataBuilder = DocDataBuilder
+  type HedlineBuilder = HeadlineBuilder
+  type PBuilder = ParagraphBuilder
+  type PubdataBuilder = PublicationDataBuilder
+
+  // aliases for enumeration types whose names are unintuitive
+  type NoteType = TypeType2
+  type PublicationType = TypeType
+  type TaglineType = TypeType3
+
+  val BareNitfNamespace = toScope(None -> defaultScope.uri)
+
+  private[builders] val DateTimeAttributeKey = "norm"
+}
+
+object Builder {
+  import scala.language.implicitConversions
+  implicit def buildBuilder[T](builder: Builder[T]): T = builder.build
+}
+
+trait Builder[T] {
+  def build: T
+
+  override def toString: String = build.toString
+}
+
+trait EnrichedTextBuilder {
+  def withText(x: String): this.type = withContent(dr(x))
+  def withXml(x: NodeSeq): this.type = withContent(dr(x))
+  def withAnchor(x: A): this.type = withEnrichedText(x)
+  def withChron(x: Chron): this.type = withEnrichedText(x)
+  def withEmphasis(x: Em): this.type = withEnrichedText(x)
+  def withEvent(x: Event): this.type = withEnrichedText(x)
+  def withMoney(x: Money): this.type = withEnrichedText(x)
+  def withNumeric(x: Num): this.type = withEnrichedText(x)
+  def withQuotation(x: Q): this.type = withEnrichedText(x)
+  def withLineBreak(x: Br): this.type = withEnrichedText(x)
+  def withLanguage(x: Lang): this.type = withEnrichedText(x)
+  def withPerson(x: Person): this.type = withEnrichedText(x)
+  def withOrganization(x: Org): this.type = withEnrichedText(x)
+  def withLocation(x: Location): this.type = withEnrichedText(x)
+  def withCopyright(x: Copyrite): this.type = withEnrichedText(x)
+  def withClassifier(x: Classifier): this.type = withEnrichedText(x)
+  def withFunction(x: FunctionType): this.type = withEnrichedText(x)
+  def withPostalAddress(x: Postaddr): this.type = withEnrichedText(x)
+  def withObjectTitle(x: ObjectTitle): this.type = withEnrichedText(x)
+  def withPronunciation(x: Pronounce): this.type = withEnrichedText(x)
+  def withVirtualLocation(x: Virtloc): this.type = withEnrichedText(x)
+
+  protected def withEnrichedText[T <: EnrichedTextOption : CanWriteXML](x: T): this.type = withContent(dr(x))
+  protected def withContent(x: DataRecord[_]): this.type
+}
+
+class NitfBuilder(var build: Nitf = Nitf(body = Body())) extends Builder[Nitf] {
```

Comment:
Is there a way we could make this `var` private? I would not want mutation to leak out of a `Builder` abstraction.

Reply:
We can have a `private var` plus a public accessor. However, this seemed a bit excessive for such simple functionality. I think of the builder as a mutator itself, so I don't see the issue with allowing the user to update its value at any point. Additionally, these builders don't provide setters for all the fields defined on these classes, and so it felt ok to let the user modify the instance in custom ways.

That said, this is just my view. How do you feel about it?

Reply:
Is there a use-case where two builders would share another one? If that is the case, then a mutation on one side would leak into the other. I don't know if that's a valid example though, by looking at your tests it seems they are ephemeral.

I only know of one similar pattern in our codebase and this is [the queries ADT](https://github.com/guardian/content-api-scala-client/blob/master/client/src/main/scala/com.gu.contentapi.client/model/Queries.scala) in our CAPI client. Admittedly the code is fairly boilerplate-y.

Reply:
Builders don't share builders. Builders only accept model instances as parameters, and each of those is an immutable `case class`. Therefore, mutating a builder cannot have an effect on another one.
(Passing builders to other builders results in an implicit extraction of the model object.)

It's possible to define a builder as a `case class`. We would have to live with nested calls to `copy` (i.e. `copy(build.copy(...))` _and_ the user would have to use a `var` (or multiple `val`) if they don't have all the data at once.

I'm not strongly opinionated regarding this matter but I feel that the mutable form provides a tiny bit of advantage. For example, in this extract from another project, we make use of mutability to update multiple builders when an optional value is set:
```scala
val docData = new DocDataBuilder().withDocId(new DocIdBuilder().withId(content.id))
val pubData = new PublicationDataBuilder().withType(Print)
content.webPublicationDate.foreach { publicationDate =>
  pubData.withPublicationDate(publicationDate)
  docData.withReleaseDate(publicationDate)
         .withIssueDate(publicationDate)
}
val nitf = new NitfBuilder().withHead(
  new HeadBuilder().withDocData(docData).withPublicationData(pubData)
).build
```